src/integrationTest/java/org/opensearch/security/privileges/ActionPrivilegesTest.java

```diff
@@ -0,0 +1,1040 @@
+            int upperBound = (int) (expectedEstimatedNumberOfBytes * 1.1);
+
+            int actualEstimatedNumberOfBytes = subject.getEstimatedStatefulIndexByteSize();
+
+            assertTrue(
+                "estimatedNumberOfBytes: " + lowerBound + " < " + actualEstimatedNumberOfBytes + " < " + upperBound,
+                lowerBound < actualEstimatedNumberOfBytes && actualEstimatedNumberOfBytes < upperBound
+            );
+        }
+
+        public StatefulIndexPrivilegesHeapSize(int numberOfIndices, int numberOfRoles, int expectedEstimatedNumberOfBytes) {
+            this.indices = createIndices(numberOfIndices);
+            this.roles = createRoles(numberOfRoles, numberOfIndices);
+            this.expectedEstimatedNumberOfBytes = expectedEstimatedNumberOfBytes;
+        }
+
+        @Parameterized.Parameters(name = "{0} indices; {1} roles; estimated number of bytes: {2}")
+        public static Collection<Object[]> params() {
+            List<Object[]> result = new ArrayList<>();
+
+            // indices; roles; expected number of bytes
+            result.add(new Object[] { 100, 10, 10_000 });
+            result.add(new Object[] { 100, 100, 13_000 });
+            result.add(new Object[] { 100, 1000, 26_000 });
+
+            result.add(new Object[] { 1000, 10, 92_000 });
+            result.add(new Object[] { 1000, 100, 94_000 });
+            result.add(new Object[] { 1000, 1000, 112_000 });
+
+            result.add(new Object[] { 10_000, 10, 890_000 });
+            result.add(new Object[] { 10_000, 100, 930_000 });
+
+            return result;
+        }
+
+        static Map<String, IndexAbstraction> createIndices(int numberOfIndices) {
+            String[] names = new String[numberOfIndices];
+
+            for (int i = 0; i < numberOfIndices; i++) {
+                names[i] = "index_" + i;
+            }
+
+            return MockIndexMetadataBuilder.indices(names).build().getIndicesLookup();
+        }
+
+        static SecurityDynamicConfiguration<RoleV7> createRoles(int numberOfRoles, int numberOfIndices) {
+            try {
+                Random random = new Random(1);
+                Map<String, Object> rolesDocument = new HashMap<>();
+                List<String> allowedActions = Arrays.asList(
+                    "indices:data/read*",
+                    "indices:admin/mappings/fields/get*",
+                    "indices:admin/resolve/index",
+                    "indices:data/write*",
+                    "indices:admin/mapping/put"
+                );
+
+                for (int i = 0; i < numberOfRoles; i++) {
+                    List<String> indexPatterns = new ArrayList<>();
+                    int numberOfIndexPatterns = Math.min(
+                        (int) ((Math.abs(random.nextGaussian() + 0.3)) * 0.5 * numberOfIndices),
+                        numberOfIndices
+                    );
+
+                    int numberOfIndexPatterns10th = numberOfIndexPatterns / 10;
```

Comment:
The table here lists size for 10k indices and 10 and 100 roles. Both are slightly below 1 MB:

https://github.com/opensearch-project/security/blob/1fef11be731e5e2a3c06c5d565faf23525ffe0d6/src/integrationTest/java/org/opensearch/security/privileges/ActionPrivilegesTest.java#L955-L965

For 1000 roles it would be also still below 1 MB - but this takes quite long to compute, that's why I left it out of the test.